[Support] Optimize signal handling file removal code

rnk opened the pull request:
clangd/clangd#1787 describes how LLVM's code for removing partially
written files when taking a fatal signal is slower than it should be.
This code was substantially rewritten by JF Bastien in
aa1333a91f8d8a060bcf5 to make it (more) signal-safe.

As written in 2018, the logic always allocates a new node for every file
we attempt to protect, and those nodes are added to a global singly
linked list and never removed. If you open a lot of output files,
suddenly output file opening becomes O(n^2), which is what happened
during clangd indexing.

Removing files on signals in a multi-threaded environment is really
complicated! We can use locks to synchronize between threads that are
writing to the list, but we cannot use locks to synchronize against
re-entrant signals, and we don't have any great tools for masking or
delaying things like SIGTERM. This makes it difficult to assert that we
have the one and only reference to a node, even after removing it from
the linked structure, so we can safely deallocate it.

This implementation sidesteps that problem by reusing nodes on the list with a null Filename pointer. The Filename is null if the file no longer needs to be removed, or if the signal handler is actively in the middle of removing that file.
* Case 1: This is the normal case, works as intended, the list should grow in length to the number of concurrently open files that must be removed on exit, which should be O(# threads)
* Case 2: In this case, the signal handler will leak the filename string without a GC root for it, which is fine, since we're exiting.

I added a benchmark, and this
change makes it 16x faster, and the profile shows that the time is no
longer mostly spent in this signal handling code.

rnk committed:
Add load test so we can test our fix

rnk committed:
[Support] Optimize signal handling file removal code

clangd/clang#1787 describes how LLVM's code for removing partially
written files when taking a fatal signal is slower than it should be.
This code was substantially rewritten by JF Bastien in
aa1333a91f8d8a060bcf5 to make it (more) signal-safe.

As written in 2018, the logic always allocates a new node for every file
we attempt to protect, and those nodes are added to a global singly
linked list and never removed. If you open a lot of output files,
suddenly output file opening becomes O(n^2), which is what happened
during clangd indexing.

Removing files on signals in a multi-threaded environment is really
complicated! We can use locks to synchronize between threads that are
writing to the list, but we cannot use locks to synchronize against
re-entrant signals, and we don't have any great tools for masking or
delaying things like SIGTERM. This makes it difficult to assert that we
have the one and only reference to a node, even after removing it from
the linked structure, so we can safely deallocate it.

This implementation sidesteps that problem by moving removed nodes onto
a recycling list. This is safe, insofar as that if a signal handler
manages to race and traverse a node that has been recycled, it will walk
over the list of recycled nodes and not delete any files. This is better
than crashing.

I added a synthetic load test in the form of a disabled gtest and this
change makes it 16x faster, and the profile shows that the time is no
longer mostly spent in this signal handling code.

rnk committed:
Merge branch 'main' into rm-file-signal-fix

rnk committed:
Merge branch 'main' into rm-file-signal-fix

rnk committed:
Tweak comment

rnk committed:
Move the load test into a benchmark

rnk committed:
Simplify benchmark

rnk committed:
sort includes and adjust comments

rnk committed:
Remove stray include

rnk committed:
Bring back the actual code change that got lost when moving the benchmark

rnk commented:
Truth in advertising: The benchmark and code were synthesized with an LLM, intended to represent clangd's actual behavior in production. The node recycling strategy was my idea. Here are some numbers demonstrating the ~16x improvement by avoiding the quadratic behavior:
https://gist.github.com/rnk/9f7fbcdf3d21e370705b64868ca3ebe4

> I looked at the code and can't convince myself that it isn't racy. My inkling is that having two lists is troublesome, and instead having a marker for "can be reused" on a node would work better?

I think this is the risky interleaving:
* add nodes to list
* take signal, acquire reference to node A
* in other thread, transfer node A from list to recycling list
* (optional) other thread allocates a new node and adds it back to the list of files to remove
* signal handler resumes, proceeds down the list, either the recycling list, or the main list

I've convinced myself that the new code will never crash because the nodes are never freed or made to point to freed nodes, but in the case where the signal handler acquires a reference to a node that is then recycled, we may not remove the rest of the files on the main list. I was willing to live with that.

Node deactivation could work, I think it's a good idea, and I will give it a try. I worry that nodes, once created, live forever, and now every output file creation operation is `O(#maxconcurrentwrites)` for the rest of the process lifetime. That's probably OK of we only do `O(threads)` concurrent writes, but I could imagine a program that opens thousands of ostreams and registers them for removal.

rnk committed:
fix formatting

rnk committed:
Merge branch 'main' into HEAD

rnk committed:
Eliminate the recycle list. I believe we can use the null filenames to track which nodes are in use

rnk committed:
Tweak some names, comments, and simplify

rnk reviewed (commented):
PTAL, this is simpler and still makes the benchmark happy. I think it's reasonable to use nullptr as the flag for node availability.

rnk committed:
Merge branch 'main' into rm-file-signal-fix

kadircet commented on llvm/lib/Support/Unix/Signals.inc:
```diff
@@ -107,8 +107,8 @@ class FileToRemoveList {
-  // Not signal-safe.
-  FileToRemoveList(const std::string &str) : Filename(strdup(str.c_str())) {}
+  // Takes ownership of \p filename.
+  FileToRemoveList(char *filename) : Filename(filename) {}
```
do we need this change? can we keep `strdup` here to make sure we don't rely on caller to provide a "long-living" pointer (or take a unique_ptr to make it explicit and release it in here)?

jfbastien replied:
Yeah I don't think this change is needed to fix the issue.

MaskRay replied:
This is needed by the reuse mechanism in `insert()`. The strdup result, `NewFilename`, must be available before the ctor call.

kadircet reviewed (commented):
oh wow this is an amazing find, thanks a lot @rnk! 

FWIW, i don't think there's anything specific to concurrent use pattern exercised by clangd in here. The main issue is existing code just keeps a list that grows throughout the full execution of a program. I guess there's only clangd and lldb with an indefinite runtime, among rest of the applications. In case you want to change benchmarks to be single threaded (but just performs some interleaved inserts/deletes)

jfbastien commented on llvm/lib/Support/Unix/Signals.inc:
Is this leak something you want to fix for clangd as well?

rnk commented on llvm/lib/Support/Unix/Signals.inc:
```diff
@@ -122,13 +122,27 @@ public:
-    while (!InsertionPoint->compare_exchange_strong(OldHead, NewHead)) {
-      InsertionPoint = &OldHead->Next;
-      OldHead = nullptr;
+    for (FileToRemoveList *Current = Head.load(); Current;
```
I don't think most applications of LLVM will concurrently open more than O(# threads) output files at once. I wouldn't want to add complexity to improve performance, I just want to avoid the growth and leak.

I think part of what makes this complicated is managing multiple linked allocations that need to have lifetimes managed. I considered having some kind of fixed-size vector and having the threads do some kind of read-copy-update pattern, but it seemed complicated. Another direction would be to try to have per-thread file lists, to avoid cache thrashing.

rnk committed:
Eliminate the possibility of leaking the filename by putting it back on the list after removing it. This may lead to double-removal if process termination is "slow", but that would manifest as system call failure.

rnk committed:
Merge branch 'main' into rm-file-signal-fix

rnk committed:
Merge remote-tracking branch 'rnk-github/rm-file-signal-fix' into rm-file-signal-fix

rnk reviewed (commented):
I think the question is: Are folks happy to approve this now, making this O(# files open) rather than O(# files every opened), or should I work to make this O(1)? I'm comfortable with the incremental change.

Another way to get O(1) would be to revert back to the list node recycling.

rnk committed:
Use a sentinel to avoid other threads inserting filenames into this list causing leaks

jfbastien commented on llvm/lib/Support/Unix/Signals.inc:
```diff
@@ -107,28 +111,43 @@ class FileToRemoveList {
+    // There are two cases where Filename can be null:
+    // - A node left behind by a previous file that we had to remove
+    // - A node whose file is actively being removed by a signal handler right
+    //   now, in which case, it's OK if this file doesn't get removed.
```
```suggestion
    // Reuse a node with a null filename (left behind by erase) if one exists.
    // There are two cases where Filename can be special:
    // - nullptr: a node left behind by a previous file that we had to remove
    // - InvalidPathSentinel: a node whose file is actively being removed by a signal handler right
    //   now, in which case, it's OK if this file doesn't get removed.
```

rnk commented on llvm/lib/Support/Unix/Signals.inc:
```diff
@@ -145,11 +164,15 @@ public:
+        // Leave an empty filename. Use CAS to avoid racing with the signal
+        // handler (which can't take the writer lock); only clear and free
+        // if we still own the pointer.
+        char *Expected = OldFilename;
+        while (!Current->Filename.compare_exchange_strong(Expected, nullptr)) {
+          if (Expected == nullptr)
+            break; // Another actor cleared it; nothing to free.
+        }
```
Makes sense, that's basically the signal handler taking ownership of the file.

rnk committed:
Apply suggestions from code review

Co-authored-by: JF Bastien <github@jfbastien.com>

rnk committed:
Merge branch 'main' into rm-file-signal-fix

rnk committed:
Merge branch 'main' into rm-file-signal-fix

rnk committed:
CMake: Include pthread in configure link checks

On glibc before 2.34, librt may depend on symbols from libpthread. LLVM's HAVE_LIBRT probe uses check_library_exists(rt shm_open ...), which links a small test against librt and CMAKE_REQUIRED_LIBRARIES. If pthread is not present in that ambient link context, the probe can fail even though shm_open is available and LLVM targets later link with pthread.

Finish pthread discovery before the dlopen and rt library checks, and add the discovered thread library to CMAKE_REQUIRED_LIBRARIES when one is needed. This makes subsequent configure checks match the link context used by LLVM's Unix targets.

The change is low risk for modern C libraries. glibc 2.34 and newer integrated pthread-related functionality into libc, so CMAKE_THREAD_LIBS_INIT is typically empty. musl likewise provides pthread functionality from libc with compatibility libraries for traditional link options. For older glibc systems, adding pthread is the required dependency for probing librt and matches the final link dependencies LLVM already uses.

rnk committed:
Revert "CMake: Include pthread in configure link checks"

This reverts commit 28142695b0356b7d036b828893e3f3320e3fe242.

This got split into PR #203054

rnk committed:
[Support] Restore empty signal removal slots

rnk commented on llvm/lib/Support/Unix/Signals.inc:
```diff
@@ -175,20 +198,18 @@ public:
-        // We're done removing the file, erasing can safely proceed.
+        // Add the path back to the list to create a global root referencing the
+        // heap allocation, which will pacify leak checkers that run at exit.
         currentFile->Filename.exchange(path);
```
It took a while, but I actually ran the stress test as you suggested, and confirmed this change resolves the issue: https://github.com/rnk/llvm-project/commit/ef1196e64ecd7c8153b683d87102af051ca50f5c

The fix I pushed is slightly different from yours in the case where we encounter a sentinel on the list, in which case, the removal is owned by another signal executing concurrently, and we do nothing, which seems right.

rnk reviewed (commented):
I got distracted, but I finally came back around to this.

rnk committed:
Merge branch 'main' into rm-file-signal-fix

rnk committed:
Merge main / rm-file-signal
